File: frontend/services/authService.ts
import api, { getToken } from "./api";
import type { LoginRequest, LoginResponse, User } from "@/types/auth";

type ApiResponse<T> = {
  status: string;
  message: string;
  data: T;
};

export async function login(payload: LoginRequest) {
  const response = await api.post<ApiResponse<LoginResponse>>("/auth/login", payload);
  return response.data.data;
}

export async function register(payload: {
  email: string;
  full_name: string;
  password: string;
  role: "admin" | "guru_bk" | "wali_kelas" | "kesiswaan" | "kepala_sekolah";
}) {
  const response = await api.post<ApiResponse<Record<string, unknown>>>("/auth/register", payload);
  return response.data.data;
}

export async function getCurrentUser() {
  const token = getToken();
  if (!token) return null;

  const response = await api.get<ApiResponse<User>>("/auth/me");
  return response.data.data;
}

export function logout() {
  if (typeof window === "undefined") return;
  localStorage.removeItem("token");
  localStorage.removeItem("user");
}

File: backend/app/services/auth_service.py
from datetime import datetime
from typing import Optional
from uuid import uuid4

from fastapi import HTTPException, status
from jose import JWTError, jwt
from sqlalchemy.orm import Session

from app.core.config import settings
from app.core.jwt import create_access_token
from app.core.security import hash_password, verify_password
from app.models.user_model import User
from app.schemas.auth_schema import UserCreate, UserLogin


ALLOWED_ROLES = {
    "admin",
    "guru_bk",
    "wali_kelas",
    "kesiswaan",
    "kepala_sekolah",
}


def get_user_by_email(db: Session, email: str) -> Optional[User]:
    return db.query(User).filter(User.email == email).first()


def create_user(db: Session, payload: UserCreate) -> User:
    if payload.role not in ALLOWED_ROLES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Role tidak valid",
        )

    existing_user = get_user_by_email(db, payload.email)
    if existing_user:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Email sudah terdaftar",
        )

    user = User(
        id=str(uuid4()),
        email=payload.email.lower(),
        full_name=payload.full_name.strip(),
        role=payload.role,
        status="aktif",
        hashed_password=hash_password(payload.password),
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow(),
    )
    db.add(user)
    db.commit()
    db.refresh(user)
    return user


def authenticate_user(db: Session, payload: UserLogin) -> User:
    user = get_user_by_email(db, payload.email.lower())
    if not user or not verify_password(payload.password, user.hashed_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Email atau password salah",
        )
    return user


def build_access_token(user: User) -> str:
    return create_access_token(
        data={
            "sub": user.id,
            "email": user.email,
            "role": user.role,
            "full_name": user.full_name,
        }
    )


def get_current_user_from_token(db: Session, token: str) -> User:
    try:
        payload = jwt.decode(
            token,
            settings.JWT_SECRET_KEY,
            algorithms=[settings.JWT_ALGORITHM],
        )
        user_id = payload.get("sub")
        if not user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token tidak valid",
            )
    except JWTError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Token tidak valid",
        )

    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="User tidak ditemukan",
        )
    return user
